### ISSUE TITLE
Route entries get deleted by felix due to stale data from typha

### ISSUE THREAD

**Issue by DockToFuture** [projectcalico/calico#5173]
## Expected Behavior

In a kubernetes cluster using calico (v3.21.1), pod network works so that pods can communicate with each other and the outside world according to network policies.

## Current Behavior

On some nodes the communication from within a pod stops working since the route from node to pod gets removed. In this case for `10.245.61.13/32`. The interface (`cali1037a54e65e`) however still exists. 

```
2021-12-10 14:10:42.201 [INFO][22936] felix/summary.go 100: Summarising 11 dataplane reconciliation loops over 1m3.3s: avg=37ms longest=316ms ()
2021-12-10 14:10:51.442 [INFO][106] monitor-addresses/startup.go 741: Using autodetected IPv4 address 10.250.15.187/22 on interface ens192 matching cidrs [10.250.12.0/22]
2021-12-10 14:11:31.653 [INFO][22936] felix/int_dataplane.go 1071: Linux interface addrs changed. addrs=set.mapSet{} ifaceName="cali1037a54e65e"
2021-12-10 14:11:31.653 [INFO][22936] felix/int_dataplane.go 1036: Linux interface state changed. ifIndex=751 ifaceName="cali1037a54e65e" state="up"
2021-12-10 14:11:31.653 [INFO][22936] felix/int_dataplane.go 1539: Received interface addresses update msg=&intdataplane.ifaceAddrsUpdate{Name:"cali1037a54e65e", Addrs:set.mapSet{}}
2021-12-10 14:11:31.653 [INFO][22936] felix/hostip_mgr.go 85: Interface addrs changed. update=&intdataplane.ifaceAddrsUpdate{Name:"cali1037a54e65e", Addrs:set.mapSet{}}
2021-12-10 14:11:31.654 [INFO][22936] felix/int_dataplane.go 1521: Received interface update msg=&intdataplane.ifaceUpdate{Name:"cali1037a54e65e", State:"up", Index:751}
2021-12-10 14:11:31.654 [INFO][22936] felix/endpoint_mgr.go 366: Workload interface came up, marking for reconfiguration. ifaceName="cali1037a54e65e"
2021-12-10 14:11:31.654 [INFO][22936] felix/endpoint_mgr.go 1179: Applying /proc/sys configuration to interface. ifaceName="cali1037a54e65e"
2021-12-10 14:11:31.655 [INFO][22936] felix/route_table.go 892: Syncing routes: found unexpected route; ignoring due to grace period. dest=10.245.61.13/32 ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 tableIndex=0
2021-12-10 14:11:31.656 [INFO][22936] felix/route_table.go 892: Syncing routes: found unexpected route; ignoring due to grace period. dest=10.245.61.13/32 ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 tableIndex=0
2021-12-10 14:11:31.656 [INFO][22936] felix/route_table.go 567: Interface in cleanup grace period, will retry after. ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 tableIndex=0
2021-12-10 14:11:33.095 [INFO][22936] felix/iface_monitor.go 201: Netlink address update. addr="fe80::ecee:eeff:feee:eeee" exists=true ifIndex=751
2021-12-10 14:11:33.095 [INFO][22936] felix/int_dataplane.go 1071: Linux interface addrs changed. addrs=set.mapSet{"fe80::ecee:eeff:feee:eeee":set.empty{}} ifaceName="cali1037a54e65e"
2021-12-10 14:11:33.095 [INFO][22936] felix/int_dataplane.go 1539: Received interface addresses update msg=&intdataplane.ifaceAddrsUpdate{Name:"cali1037a54e65e", Addrs:set.mapSet{"fe80::ecee:eeff:feee:eeee":set.empty{}}}
2021-12-10 14:11:33.095 [INFO][22936] felix/hostip_mgr.go 85: Interface addrs changed. update=&intdataplane.ifaceAddrsUpdate{Name:"cali1037a54e65e", Addrs:set.mapSet{"fe80::ecee:eeff:feee:eeee":set.empty{}}}
2021-12-10 14:11:33.096 [INFO][22936] felix/route_table.go 892: Syncing routes: found unexpected route; ignoring due to grace period. dest=10.245.61.13/32 ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 tableIndex=0
2021-12-10 14:11:33.097 [INFO][22936] felix/route_table.go 892: Syncing routes: found unexpected route; ignoring due to grace period. dest=10.245.61.13/32 ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 tableIndex=0
2021-12-10 14:11:33.097 [INFO][22936] felix/route_table.go 567: Interface in cleanup grace period, will retry after. ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 tableIndex=0
2021-12-10 14:11:35.029 [INFO][22936] felix/route_table.go 892: Syncing routes: found unexpected route; ignoring due to grace period. dest=10.245.61.13/32 ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 tableIndex=0
2021-12-10 14:11:35.030 [INFO][22936] felix/route_table.go 892: Syncing routes: found unexpected route; ignoring due to grace period. dest=10.245.61.13/32 ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 tableIndex=0
2021-12-10 14:11:35.031 [INFO][22936] felix/route_table.go 567: Interface in cleanup grace period, will retry after. ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 tableIndex=0
2021-12-10 14:11:39.125 [INFO][22936] felix/route_table.go 892: Syncing routes: found unexpected route; ignoring due to grace period. dest=10.245.61.13/32 ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 tableIndex=0
2021-12-10 14:11:39.126 [INFO][22936] felix/route_table.go 892: Syncing routes: found unexpected route; ignoring due to grace period. dest=10.245.61.13/32 ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 tableIndex=0
2021-12-10 14:11:39.126 [INFO][22936] felix/route_table.go 567: Interface in cleanup grace period, will retry after. ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 tableIndex=0
2021-12-10 14:11:45.779 [INFO][22936] felix/route_table.go 896: Remove old route dest=10.245.61.13/32 ifaceName="cali1037a54e65e" ifaceRegex="^cali.*" ipVersion=0x4 routeProblems=[]string{"unexpected route"} tableIndex=0
2021-12-10 14:11:45.779 [INFO][22936] felix/conntrack.go 90: Removing conntrack flows ip=10.245.61.13
2021-12-10 14:11:45.782 [INFO][22936] felix/summary.go 100: Summarising 13 dataplane reconciliation loops over 1m3.6s: avg=72ms longest=703ms ()
2021-12-10 14:11:51.465 [INFO][106] monitor-addresses/startup.go 741: Using autodetected IPv4 address 10.250.15.187/22 on interface ens192 matching cidrs [10.250.12.0/22]
```

As consequence of the deleted routes we saw in the kernel ring buffer the following martian packets(data from another cluster):
```
[3643049.978035] IPv4: martian source 169.254.1.1 from 100.96.5.62, on dev cali305be8838b5
[3643049.984308] ll header: 00000000: ff ff ff ff ff ff 3a cd 6b 6b 10 ff 08 06
[3643052.927980] IPv4: martian source 169.254.1.1 from 100.96.5.62, on dev cali305be8838b5
[3643052.934388] ll header: 00000000: ff ff ff ff ff ff 3a cd 6b 6b 10 ff 08 06
[3643053.946057] IPv4: martian source 169.254.1.1 from 100.96.5.62, on dev cali305be8838b5
[3643053.952332] ll header: 00000000: ff ff ff ff ff ff 3a cd 6b 6b 10 ff 08 06
[3643054.970021] IPv4: martian source 169.254.1.1 from 100.96.5.62, on dev cali305be8838b5
[3643054.976333] ll header: 00000000: ff ff ff ff ff ff 3a cd 6b 6b 10 ff 08 06
[3643057.928153] IPv4: martian source 169.254.1.1 from 100.96.5.62, on dev cali305be8838b5
[3643057.934515] ll header: 00000000: ff ff ff ff ff ff 3a cd 6b 6b 10 ff 08 06
[3643058.937977] IPv4: martian source 169.254.1.1 from 100.96.5.62, on dev cali305be8838b5
[3643058.944272] ll header: 00000000: ff ff ff ff ff ff 3a cd 6b 6b 10 ff 08 06
[3643059.961983] IPv4: martian source 169.254.1.1 from 100.96.5.62, on dev cali305be8838b5
[3643060.062278] ll header: 00000000: ff ff ff ff ff ff 3a cd 6b 6b 10 ff 08 06
```
Some of the existing pods on a failing node are effected. New pods on a failing node are always effected.

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Context

We use typha with multiple instances, in this case two. All `calico-nodes` with issues were connected to one typha instance while the rest of the cluster seems to work as expected. A restart of the "stale" typha resolved the problems.
Reconfiguring felix through a change to the `felixconfiguration` seems to trigger a reconnect to typha. Using this we could toggle between the "stale" and working typha instance.

The issue was first observed after updating to calico `v3.21.x`, `v3.19.x` and `v3.20.x` work as expected.

## Your Environment

* Calico version v3.21.1
* Orchestrator version (e.g. kubernetes, mesos, rkt): kubernetes
* Operating System and version: 

/cc: @scheererj

**Comment by fasaxc** [comment]
This kind of issue is typically caused by a bug in the kubernetes client.  v3.20.x used kube client v0.21.0-rc.0 but v3.21 uses v0.21.0.  Perhaps there was a regression.

**Comment by cdtomkins** [comment]
@DockToFuture Thanks for raising this. We would like to gather additional information. Specifically, can you confirm:
- the current K8s version and any other versions you have observed this on
- details on the K8s environment - particularly the build process and anything unusual or custom, such as extra networking hops, load balancers, or anything like that.

**Comment by DockToFuture** [comment]
The k8s versions we observed the problem on were `1.19.15`, `1.20.9` and `1.22.2`.
We use https://github.com/gardener/gardener for creating and managing our k8s clusters.
The underlying infrastructures were openstack, aws and azure. 
The controlplane is in a different infrastructure account than the dataplane, see https://gardener.cloud for more details. Between the dataplane and the kube-apiserver is a public loadbalancer and an istio ingress gateway. The traffic going via the `kubernetes.default.svc.cluster.local` service goes over an additional envoy proxy.

**Comment by fasaxc** [comment]
Eek, that's a lot of moving parts between Calico and your API server.  Calico relies on HTTP2 keepalives to make sure the connection to the API server is up.  I wonder if one of your intermediate proxies is preventing keepalives from flowing end-to-end.

**Comment by ScheererJ** [comment]
This setup has worked for more than a year like that with different calico versions. Only with calico v3.21.x we sometimes see the issue described here.

**Comment by fasaxc** [comment]
I've tried to reproduce the issue by simulating a silent failure of the API server connection but in my tests, the failure is detected as expected.  Are you sure that only Calico changed to cause the issue?  Do different versions of Gardener also come with different versions of Envoy/Istio/etc?

**Comment by DockToFuture** [comment]
Currently we use the same istio and envoy version for all clusters and it works smoothly with the old/prior calico versions.

**Comment by djsly** [comment]
AKS started deploying v3.21.0 today and on each clusters where the new version was deployed, we started seeing the same issue, routes were removed, we even saw Felix trying to reach out to typha pods that didn't exists anymore. (wrong IP).  Due to the impact it had, it was reverted back to v3.20.0 and it's now stable, we didn't keep around the calico-node logs. I would try to dig in our logging backend to see if we are indexing the calico-node logs.

**Comment by caseydavenport** [comment]
I'd like to look into this, but probably need a full log file from a calico/typha + calico/node that are experiencing this issue, and if possible a /var/log/calico/cni/cni.log from the node as well (assuming it is using calico CNI).

This is the full set of changes from v3.20.0 -> v3.21.2 in the calico/typha codebase: https://github.com/projectcalico/typha/compare/v3.20.0...v3.21.2

There's not much there, but a few things stand out as potential disruptions:

 - https://github.com/projectcalico/typha/pull/614/
 - https://github.com/projectcalico/typha/pull/694
 - https://github.com/projectcalico/typha/pull/671

It may have also been a change in libcalico-go though, as much of connection logic to the apiserver lives there. 

These are the changes in libcalico between typha v3.20.0 and v3.21.2: https://github.com/projectcalico/libcalico-go/compare/279457908a6...3402eca9d27

The main changes that stand out to me are:
- https://github.com/projectcalico/libcalico-go/pull/1484


Some other thoughts:

> AKS started deploying v3.21.0 today

v3.21.0 had a bug fixed in v3.21.1 which impacted pods that made use of Kubernetes hostPorts - if the pods that were losing connectivity all had hostPorts (specifically without having a name set) then it is likely you were hitting the issue fixed here: https://github.com/projectcalico/libcalico-go/pull/1545

**Comment by fasaxc** [comment]
> Felix gets this information by direct query of the Kubernetes Endpoints API. Maybe this is the apiserver telling Felix the wrong information? Or Maybe the Typha pod goes away and Felix never realizes it, and so never re-queries the apiserver for a new endpoint?

Casey is right, when Felix starts up, it queries the API server directly for the list of service endpoints

The Felix/typha prototcol includes a layer-7 heartbeat.  If Felix doesn't hear from typha for 30s then it will disconnect and restart.  Then it goes through discovery again, reading the list of typha endpoints again and choosing a fresh one.

**Comment by fasaxc** [comment]
@DockToFuture @djsly When you say that you're rolling out Calico v3.21.x, is that the _only_ change you're making on these clusters?  Can you give us a timeline of events?  You apply the new v3.21.x update and then how long until you notice stale routes?

**Comment by DockToFuture** [comment]
Hello @fasaxc, once we rolled it out it took a while until we noticed the first issues with that problem. We saw it sporadically on a few clusters but not on the majority of clusters.
We reverted our calico version back to the previous one prior v3.21.x where we faced no issues.

**Comment by fasaxc** [comment]
Thanks @DockToFuture, are we talking 1-in-100 clusters, 1-in-10?  By "while" are we talking 15 mins, or, you upgraded one day and then found issues a day later (for example)?

Anything that stands out about the affected clusters?  Were they small clusters; large clusters; high churn rate of pods (or the opposite)?

**Comment by DockToFuture** [comment]
We are in the range of ~10 out of 1000 clusters where we noticed it. We saw the issue on small and on large clusters. I would say it was evenly distributed. We noticed the issue after a couple of days, but since we also using a hibernation feature on many clusters, the issue gets resolved on its own due to a restart of typha. So there is a lot of uncertainty in the observations. I can not tell if the issue occurred directly after the rollout of the new calico version or if it took some time for typha to get in that "stale" state.

**Comment by djsly** [comment]
for me, AKS started the upgrade in the background, we saw the issue on 4 / 4 of the upgraded clusters, It seems that for the first cluster, nothing was failing for 24hours until we started deploying new pods (scalling up a deployments). So post calico upgrade, nothing seems to fail right away, but once we started performing operations, thats when users started reporting issues. 

For us, it was ALL clusters that were affected post calico upgrades, we are running AKS 1.20.9 with both linux and windows node pools , one cluster has 50 nodes with 1000 pods while the other has  over 400 nodes and ~7500 pods

**Comment by fasaxc** [comment]
OK, so that is consistent with typha starting, connecting to the datastore, downloading a snapshot and generally working for a while.  Then, at some point, the connection to the datastore stalls and typha gets stuck.  Since typha is a caching proxy, it'll cache the most recent state of the datastore that it has seen and that's what felix will see if it connects to _that_ typha.

**Comment by fasaxc** [comment]
We're struggling to reproduce this issue.  We've made cluster with hundreds of typha instances and using their prometheus metrics to look for any laggards but so far no luck.

@DockToFuture @ffilippopoulos  would you be able to try test builds of typha with extra diagnostics and some of the candidate causes backed out?

**Comment by djsly** [comment]
have you tried putting extra stress on the apiserver ? causing watch cache connections to drop / die/ reconnect.  that would be my guess

**Comment by djsly** [comment]
one of the cluster, the issue started ~2h after, when the deployments started, let me try to find some time today to extra the logs, you wanted Typha and Calico Node ?

**Comment by fasaxc** [comment]
Yes please, and any timeline information about when exactly you saw what problems would be useful.

**Comment by djsly** [comment]
here are logs from 3:30 UTM to 4:30 UTM, users reported seeing login failure at 4:00UTC

logs are from ALL calico pods (typha/node(s))

[query_data-5.csv](https://github.com/projectcalico/calico/files/7757962/query_data-5.csv)

**Comment by djsly** [comment]
Logs standing out 

```
2021-12-17 15:27:07.984 [INFO][6] watchercache.go 188: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/workloadendpoints" error=ListOptions.meta.k8s.io "" is invalid: [resourceVersionMatch: Forbidden: resourceVersionMatch is forbidden unless resourceVersion is provided, resourceVersionMatch: Forbidden: resourceVersionMatch is forbidden when continue is provided]
```

**Comment by caseydavenport** [comment]
Hmm, I think that might be a bug in the Kubernetes client, based on a quick investigation. 

That error is coming from here: https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/internalversion/validation/validation.go#L35-L37

The Calico code uses the client-go list pager to perform paged List() calls against the Kubernetes API here, passing in resourceVersion as well as resourceVersionMatch as needed: https://github.com/projectcalico/calico/blob/8fdb4215535b71117917bfa8871abe2652ec3a25/libcalico-go/lib/backend/k8s/resources/list_pager.go#L43-L50

The client-go code properly clears resourceVersion on page continuations, but doens't clear resourceVersionMatch: https://github.com/kubernetes/client-go/blob/master/tools/pager/pager.go#L138-L141

I wonder if that is intentional, if so I don't see a way for us to set resourceVersionMatch on paged list calls, so I would guess it is  not.

**Comment by caseydavenport** [comment]
I've raised the following issue to see what the client-go folks say: https://github.com/kubernetes/client-go/issues/1042

**Comment by ffilippopoulos** [comment]
> would you be able to try test builds of typha with extra diagnostics and some of the candidate causes backed out?

sure, we could try this on lower environments here, and keep an eye. Even though we have only seen the issue surfacing on our prod cluster so far for some reason.

**Comment by fasaxc** [comment]
I've built a test image with @caseydavenport 's suggested patch to the Kubernetes client library and pushed as docker image: `fasaxc/typha:v3.21-with-pager-fix`  This should be a drop-in replacement for `calico/typha:v3.21.x` (on AMD64 only) if you want to try it.  I was able to write a unit test that showed the above error before applying the fix and the error cleared when I applied the fix.  

I'm pretty sure the fix is correct; the reason for blocking the revision-related parameters when doing a paged read is that the pagination information is stored server side and it wouldn't make sense to change the revision on subsequent paged reads.

**Comment by fbsb** [comment]
Hi, we also had some issues when upgrading to v3.21.1 and had to revert to v3.20 which after some investigating it turned out to be caused by the same issue described here. 

It seems the fix didn't make it to v3.21.2 and were wondering if it will be included in v3.21.3 release.

**Comment by fasaxc** [comment]
Yes, it'll be in v3.21.3; it was backported here: https://github.com/projectcalico/calico/pull/5332.
